### PR TITLE
UCS/DEBUG: Add an option to disable signal traps set in UCS.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
Add an option to disable signal traps set in UCS.

## Why ?
Needed for managed languages like java, that set their own signal handlers.

## How ?
In `ucs_set_signal_handler` when setting new signal handler - save previous signal handler into hash map.
When requested restore signal handler - pop from map original signal handler and set it.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9376/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9377/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9396/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9405/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
Strange locally all valgrind tests passed:
```
[----------] Global test environment tear-down
[==========] 8314 tests from 570 test cases ran. (10058573 ms total)
[  PASSED  ] 8314 tests.
```

**Comment by petro-rudenko** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9430/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9435/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
Ok know where's an issue. Will fix.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9475/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9476/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9480/ for details (Mellanox internal link).

File: src/ucs/debug/debug.c [hunk 1]
```diff
@@ -28,6 +28,7 @@
 
 
 KHASH_MAP_INIT_INT64(ucs_debug_symbol, char*);
+KHASH_MAP_INIT_INT(ucs_signal_restorer, struct sigaction*);
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: rename: ucs_signal_orig_action

File: src/ucs/debug/debug.c [hunk 2]
```diff
@@ -988,11 +990,18 @@ void ucs_handle_error(const char *message)
 static int ucs_debug_is_error_signal(int signum)
 {
     int i;
+    khiter_t hash_it;
 
     if (!ucs_global_opts.handle_errors) {
         return 0;
     }
 
+    // If this signal is error, but was disabled.
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: C-style comment

File: src/ucs/debug/debug.c [hunk 3]
```diff
@@ -988,11 +990,18 @@ void ucs_handle_error(const char *message)
...
+    hash_it = kh_get(ucs_signal_restorer, &ucs_signal_restorer_map, signum);
+    if (hash_it == kh_end(&ucs_signal_restorer_map)) {
+        return 0;
+    }
+
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: no need to check ucs_global_opts.error_signals now

File: src/ucs/debug/debug.c [hunk 4]
```diff
@@ -1111,6 +1122,19 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
                      ucs_global_opts.error_signals.signals[i]);
         }
         ucs_debug_signal_restorer = old_action.sa_restorer;
+
+        hash_it = kh_get(ucs_signal_restorer,
+                         &ucs_signal_restorer_map,
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: add to prev line

File: src/ucs/debug/debug.c [hunk 5]
```diff
@@ -1111,6 +1122,19 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
...
+                         ucs_global_opts.error_signals.signals[i]);
+        if (hash_it == kh_end(&ucs_signal_restorer_map)) {
+            struct sigaction *oact_copy = (struct sigaction*)ucs_malloc(sizeof(old_action), "signal restorer");
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: define  struct sigaction *oact_copy in beginning of the function

File: src/ucs/debug/debug.c [hunk 6]
```diff
@@ -1111,6 +1122,19 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
...
+            memcpy(oact_copy, &old_action, sizeof(old_action));
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: can use '=' instead of memcpy

File: src/ucs/debug/debug.c [hunk 7]
```diff
@@ -1188,15 +1213,50 @@ void ucs_debug_init()
 void ucs_debug_cleanup(int on_error)
 {
     char *sym;
+    int signum;
+    struct sigaction *hndl;
 
     if (ucs_global_opts.handle_errors) {
-        ucs_set_signal_handler(NULL);
+        kh_foreach_key(&ucs_signal_restorer_map, signum,
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: could remove 'if (ucs_global_opts.handle_errors) {'

File: src/ucs/debug/debug.c [hunk 8]
```diff
@@ -1090,6 +1099,8 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
     struct sigaction sigact, old_action;
     int i;
     int ret;
+    int hash_extra_status;
+    khiter_t hash_it;
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: 'if (handler == NULL) {' could be removed now

File: src/ucs/debug/debug.c [hunk 9]
```diff
@@ -1188,15 +1213,50 @@ void ucs_debug_init()
...
+        ucs_warn("ucs_debug_disable_signal: signal %d was not set in ucs",
+                 signum);
+        return;
+    }
+    original_action = kh_val(&ucs_signal_restorer_map, hash_it);
+    ret = orig_sigaction(signum, original_action, &ucs_action);
+    if (ret < 0) {
+        ucs_warn("failed to set signal handler for sig %d : %m", signum);
+    }
+    kh_del(ucs_signal_restorer, &ucs_signal_restorer_map, hash_it);
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: space line before

File: src/ucs/debug/debug.c [hunk 10]
```diff
@@ -1188,15 +1213,50 @@ void ucs_debug_init()
...
+    int ret;
+    hash_it = kh_get(ucs_signal_restorer,
+                     &ucs_signal_restorer_map,
+                     signum);
+    if (hash_it == kh_end(&ucs_signal_restorer_map)) {
+        ucs_warn("ucs_debug_disable_signal: signal %d was not set in ucs",
+                 signum);
+        return;
+    }
+    original_action = kh_val(&ucs_signal_restorer_map, hash_it);
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: space line before

File: src/ucs/debug/debug.c [hunk 11]
```diff
@@ -1090,6 +1099,8 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
     struct sigaction sigact, old_action;
     int i;
     int ret;
+    int hash_extra_status;
+    khiter_t hash_it;
 
     if (handler == NULL) {
         sigact.sa_handler   = SIG_DFL;
```
Comment (petro-rudenko) [inline src/ucs/debug/debug.c]: Can clean this.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9511/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9518/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9525/ for details (Mellanox internal link).

File: src/ucs/debug/debug.c [hunk 12]
```diff
@@ -1188,15 +1209,46 @@ void ucs_debug_init()
...
+        kh_destroy_inplace(ucs_signal_orig_action, &ucs_signal_orig_action_map);
+    }
+}
+
+void ucs_debug_disable_signal(int signum)
+{
+    khiter_t hash_it;
+    struct sigaction *original_action, ucs_action;
+    int ret;
+    hash_it = kh_get(ucs_signal_orig_action, &ucs_signal_orig_action_map,
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: space line before

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9532/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9539/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9540/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9544/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9549/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@petro-rudenko don't force push please, i can't see what has been changes since last review.

File: src/ucs/debug/debug.c [hunk 13]
```diff
@@ -987,19 +991,15 @@ void ucs_handle_error(const char *message)
...
 
-    for (i = 0; i < ucs_global_opts.error_signals.count; ++i) {
-        if (signum == ucs_global_opts.error_signals.signals[i]) {
-            return 1;
-        }
-    }
-
-    return 0;
+    /* If this signal is error, but was disabled. */
+    hash_it = kh_get(ucs_signal_orig_action, &ucs_signal_orig_action_map, signum);
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: lock?

File: src/ucs/debug/debug.c [hunk 14]
```diff
@@ -1188,15 +1213,52 @@ void ucs_debug_init()
...
+    ucs_free(original_action);
+
+out:
+    pthread_spin_unlock(&ucs_kh_lock);
+}
+
+void ucs_debug_disable_signals()
+{
+    int signum;
+    kh_foreach_key(&ucs_signal_orig_action_map, signum,
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: lock?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9586/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
bot:mlx:retest

File: src/ucs/debug/debug.c [hunk 15]
```diff
@@ -1191,9 +1195,17 @@ unsigned long ucs_debug_get_lib_base_addr()
...
+
     kh_init_inplace(ucs_debug_symbol, &ucs_debug_symbols_cache);
     kh_init_inplace(ucs_signal_orig_action, &ucs_signal_orig_action_map);
-    pthread_spin_init(&ucs_kh_lock, 0);
+    pthread_mutexattr_init(&attr);
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+    ret = pthread_mutex_init(&ucs_kh_lock, &attr);
+    if (ret != 0) {
+         ucs_error("failed to initialize recursive lock: %s", strerror(ret));
+    }
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: space line after

File: src/ucs/debug/debug.c [hunk 16]
```diff
@@ -1253,12 +1265,14 @@ void ucs_debug_disable_signal(int signum)
...
 
 out:
-    pthread_spin_unlock(&ucs_kh_lock);
+    pthread_mutex_unlock(&ucs_kh_lock);
 }
 
 void ucs_debug_disable_signals()
 {
     int signum;
+    pthread_mutex_lock(&ucs_kh_lock);
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: space line before

File: src/ucs/debug/debug.c [hunk 17]
```diff
@@ -116,7 +116,7 @@ static stack_t  ucs_debug_signal_stack    = {NULL, 0, 0};
 khash_t(ucs_debug_symbol) ucs_debug_symbols_cache;
 khash_t(ucs_signal_orig_action) ucs_signal_orig_action_map;
 
-static pthread_spinlock_t ucs_kh_lock;
+static pthread_mutex_t ucs_kh_lock;
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: let's use ucs_spinlock_t, it's also recursive, and more lightweight

File: src/ucs/debug/debug.c [hunk 18]
```diff
@@ -1188,15 +1225,54 @@ void ucs_debug_init()
...
+out:
+    pthread_mutex_unlock(&ucs_kh_lock);
+}
+
+void ucs_debug_disable_signals()
+{
+    int signum;
+    pthread_mutex_lock(&ucs_kh_lock);
+    kh_foreach_key(&ucs_signal_orig_action_map, signum,
+                   ucs_debug_disable_signal(signum));
```
Comment (yosefe) [inline src/ucs/debug/debug.c]: is this the only reason this lock has to be recursive?
another option is to add an internal helper function like ucs_debug_disable_signal_nolock()..

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9590/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9609/ for details (Mellanox internal link).

**Comment by petro-rudenko** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9611/ for details (Mellanox internal link).

**Pull request merged**